### PR TITLE
token voting

### PR DESCRIPTION
**Pull request by johannbarbie**
resolves https://github.com/leapdao/leap-contracts/issues/261

### CODE REVIEW

File: contracts/TokenGovernance.sol [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    } else {
+      proposal.noVotes -= uint256(votes * -1);
+    }
+  }
+
+  function registerProposal(bytes32 _proposalHash) public {
+    // make sure same proposals hasn't been opened before
+    require(proposals[_proposalHash].openTime == 0, "proposal already exists");
+    // get instance of token contract and pull proposal stake
+    leapToken.transferFrom(msg.sender, address(this), PROPOSAL_STAKE);
```
Comment (MaxStalker) [inline contracts/TokenGovernance.sol]: I was wondering, what's gonna happen if initiator don't have enough LEAP? 🤔

File: contracts/TokenGovernance.sol [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    require(timestamp <= proposal.openTime, "The transaction was submitted after the vote open time");
+    
+    bytes memory txData;
+    bytes32 txHash;
+    (, txHash, txData) = TxLib.validateProof(96, _proof);
+
+    // parse tx and check if it is usable for voting
+    TxLib.Tx memory tx = TxLib.parseTx(txData);
+    TxLib.Output memory out = tx.outs[_outputIndex];
+    require(out.value > 0, "UTXO has no value");
```
Comment (MaxStalker) [inline contracts/TokenGovernance.sol]: I remember that we had a conversation, discussing if the UTXO should be of certain value, not just ANY `Good UTXO`

File: contracts/TokenGovernance.sol [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+  function challengeUTXO(bytes32 _proposalHash, bytes32[] memory _proof, uint8 _inputIndex) public {
+    uint32 timestamp;
+    (, timestamp,,) = bridge.periods(_proof[0]);
+    require(timestamp > 0, "The referenced period was not submitted to bridge");
+    require(timestamp <= proposals[_proposalHash].openTime, "The transaction was submitted after the vote open time");
+    
+    bytes memory txData;
+    bytes32 txHash;
+    (, txHash, txData) = TxLib.validateProof(64, _proof);
```
Comment (MaxStalker) [inline contracts/TokenGovernance.sol]: We are not using txHash anyware in this method... 🤷‍♂️

File: contracts/TokenGovernance.sol [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    // parse tx
+    TxLib.Tx memory txn = TxLib.parseTx(txData);
+    // checking that the transactions has been used in a vote
+    address signer = proposals[_proposalHash].usedTxns[txn.ins[_inputIndex].outpoint.hash];
+    // note: we don't check the output index here, assume that each transactions is only used for 1 vote
```
Comment (MaxStalker) [inline contracts/TokenGovernance.sol]: I believe, you are reffering here that transaction can have more than a single output and - technically - one should be able to vote with all of them separetly?

File: contracts/TokenGovernance.sol (hunk 1 above)
Comment (johannbarbie) [reply contracts/TokenGovernance.sol]: @iamonuwa and me think it will throw 😁

File: contracts/TokenGovernance.sol (hunk 4 above)
Comment (johannbarbie) [reply contracts/TokenGovernance.sol]: yes, exactly. i would like to move this to a later version. not sure it is worth covering the case now.

**Pull request merged**